scripts/benchmark_db/benchmark_base.rb

```diff
@@ -0,0 +1,263 @@
+    File.write(config_path, toml_str)
+  end
+  private :build_config_file
+
+  # Helper function for `build_artefacts` function; defines config path, snapshot
+  # path, and start epoch for building and running client.
+  def build_substitution_hash
+    height = snapshot_height(@snapshot_path)
+    start = height - @heights
+
+    return { c: '<tbd>', s: '<tbd>', h: start } if @dry_run
+
+    config_path = "#{data_dir}/#{@name}.toml"
+
+    { c: config_path, s: @snapshot_path, h: start }
+  end
+  private :build_substitution_hash
+
+  # Helper function for `build_artefacts`; clones repository, checks out branch,
+  # runs clean command, and builds client.
+  def build_client
+    if Dir.exist?(repository_name)
+      @logger.warn "Directory #{repository_name} is already present"
+    else
+      @logger.info 'Cloning repository'
+      clone_command
+      Dir.mkdir(repository_name) if @dry_run
+      @created_repository = true
+    end
+
+    @logger.info 'Clean and build client'
+    Dir.chdir(repository_name) do
+      checkout_command
+      clean_command
+      build_command
+    end
+  end
+
+  # Calls all helper functions to prepare each client to run benchmarks.
+  def build_artefacts
+    @logger.info 'Building artefacts...'
+    build_client
+
+    build_config_file unless @dry_run
+    build_substitution_hash
+  end
+  private :build_artefacts
+end
+
+# Mixin module for base benchmark class run and validation commands.
+module RunCommands
+  # Create and call proper validation command, then write results to metrics.
+  def run_validation_step(daily, args, metrics)
+    unless daily
+      validate_command = splice_args(@validate_command, args)
+      metrics[:validate] = exec_command(validate_command)
+      return
+    end
+
+    validate_online_command = splice_args(@validate_online_command, args)
+    new_metrics = exec_command(validate_online_command, self)
+    new_metrics[:tpm] =
+      new_metrics[:num_epochs] ? new_metrics[:num_epochs] / online_validation_secs : 'n/a'
+    new_metrics[:tpm] = new_metrics[:tpm].ceil(3)
+    metrics[:validate_online] = new_metrics
+  end
+
+  # Import snapshot, write metrics, and call validation function, returning metrics.
+  def import_and_validation(daily, args, metrics)
+    import_command = splice_args(@import_command, args)
+    metrics[:import] = exec_command(import_command)
+
+    # Save db size just after import.
+    metrics[:import][:db_size] = db_size unless @dry_run
+
+    run_validation_step(daily, args, metrics)
+    metrics
+  rescue StandardError, Interrupt
+    @logger.error('Fiasco during benchmark run. Deleting downloaded files, cleaning DB and stopping process...')
+    FileUtils.rm_f(@snapshot_path) if @snapshot_downloaded
+    clean_db
+    FileUtils.rm_rf(repository_name) if @created_repository
+    exit(1)
+  end
+
+  # This is the primary function called in `bench.rb` to run the metrics for
+  # each benchmark.
+  def run(daily, snapshot_downloaded)
```

Comment:
When running forest bench I had:

```
2023-05-22T09:49:16.596730Z  WARN forest_paramfetch: Error checking file: Too many open files (os error 24)    
2023-05-22T09:49:16.596717Z  WARN forest_paramfetch: Error checking file: Too many open files (os error 24)   
```

We should make sure before running the script that enough preconditions are met.
- enough free disk space
- `ulimit -n` is correctly set

Reply:
As we discussed, adding this to #2683